### PR TITLE
Refactor Dockerfile: Fix Python version; Add CI workflow to test Docker container

### PR DESCRIPTION
**Pull request by fabridamicelli**
## Description

~There is something funny going on with the non-default python version in the `pytorch/torchserve` image.
Basically the final image still uses the wrong python version (even jumping to 2.7) (despite of building with the flag -py 3.9/10).
I am tracing down the error, since there is a bit of magic going on with symlinks, PATH, etc. in the different layers and still haven't completely solved it. But for now the above mentioned typo is a clear bug. ~
I will update this when I find the rest of the problem.~  
Finally spotted the deeper error; there are two fundamental problems:
 - A typo in the Dockerfilie: PYHON_VERSION (missing Y!) --> PYTHON_VERSION.
 - The ARG PYTHON_VERSION was missing in the second build-stage. We have to redefine ARGs and ENV variables on each stage because they do not persist across them. Being empty, it silently becomes "" which does not fail, but breaks current version handling logic and resulting in images with wrong python version.
This leads to having to install python again in the last stage, which was previously not necessary because it wasn't really working.

I believe we could build a test in CI to prevent this error in the future by testing the python version explicitly in the built images.
So please let me know if there is interest to pursue that and I could work on it (maybe on a sepate issue?).

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

## Feature/Issue validation/testing
Please describe the Unit or Integration tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [x] python version in current `pytorch/torchserve:latest-cpu` 
```
~/projects/serve/docker(master)$ ./build_image.sh -py 3.10 
Valid python version
[+] Building 1.4s (23/23) FINISHED                                                                                                                                                                                                   
 => [internal] load build definition from Dockerfile                                                                                                                                                                            0.0s
 => => transferring dockerfile: 38B                                                                                                                                                                                             0.0s
 => [internal] load .dockerignore                                                                                                                                                                                               0.0s
 => => transferring context: 2B                                                                                                                                                                                                 0.0s
 => resolve image config for docker.io/docker/dockerfile:experimental                                                                                                                                                           0.4s
 => CACHED docker-image://docker.io/docker/dockerfile:experimental@sha256:600e5c62eedff338b3f7a0850beb7c05866e0ef27b2d2e8c02aa468e78496ff5                                                                                      0.0s
 => [internal] load metadata for docker.io/library/ubuntu:20.04                                                                                                                                                                 0.6s
 => [internal] load build context                                                                                                                                                                                               0.0s
 => => transferring context: 80B                                                                                                                                                                                                0.0s
 => [compile-image 1/8] FROM docker.io/library/ubuntu:20.04@sha256:24a0df437301598d1a4b62ddf59fa0ed2969150d70d748c84225e6501e9c36b9                                                                                             0.0s
 => CACHED [runtime-image 2/9] RUN --mount=type=cache,target=/var/cache/apt     apt-get update &&     DEBIAN_FRONTEND=noninteractive apt-get install --no-install-recommends -y     python$PYTHON_VERSION     python3-distutil  0.0s
 => CACHED [runtime-image 3/9] RUN useradd -m model-server     && mkdir -p /home/model-server/tmp                                                                                                                               0.0s
 => CACHED [compile-image 2/8] RUN --mount=type=cache,id=apt-dev,target=/var/cache/apt     apt-get update &&     apt-get upgrade -y &&     apt-get install software-properties-common -y &&     add-apt-repository -y ppa:dead  0.0s
 => CACHED [compile-image 3/8] RUN update-alternatives --install /usr/bin/python python /usr/bin/python3.10 1     && update-alternatives --install /usr/local/bin/pip pip /usr/local/bin/pip3 1                                 0.0s
 => CACHED [compile-image 4/8] RUN python$PYHON_VERSION -m venv /home/venv                                                                                                                                                      0.0s
 => CACHED [compile-image 5/8] RUN python -m pip install -U pip setuptools                                                                                                                                                      0.0s
 => CACHED [compile-image 6/8] RUN export USE_CUDA=1                                                                                                                                                                            0.0s
 => CACHED [compile-image 7/8] RUN TORCH_VER=$(curl --silent --location https://pypi.org/pypi/torch/json | python -c "import sys, json, pkg_resources; releases = json.load(sys.stdin)['releases']; print(sorted(releases, key  0.0s
 => CACHED [compile-image 8/8] RUN python -m pip install -U setuptools && python -m pip install --no-cache-dir captum torchtext torchserve torch-model-archiver                                                                 0.0s
 => CACHED [runtime-image 4/9] COPY --chown=model-server --from=compile-image /home/venv /home/venv                                                                                                                             0.0s
 => CACHED [runtime-image 5/9] COPY dockerd-entrypoint.sh /usr/local/bin/dockerd-entrypoint.sh                                                                                                                                  0.0s
 => CACHED [runtime-image 6/9] RUN chmod +x /usr/local/bin/dockerd-entrypoint.sh     && chown -R model-server /home/model-server                                                                                                0.0s
 => CACHED [runtime-image 7/9] COPY config.properties /home/model-server/config.properties                                                                                                                                      0.0s
 => CACHED [runtime-image 8/9] RUN mkdir /home/model-server/model-store && chown -R model-server /home/model-server/model-store                                                                                                 0.0s
 => CACHED [runtime-image 9/9] WORKDIR /home/model-server                                                                                                                                                                       0.0s
 => exporting to image                                                                                                                                                                                                          0.0s
 => => exporting layers                                                                                                                                                                                                         0.0s
 => => writing image sha256:a27b8be180bb7feeb8212d4b1f96085dbbe4ab931478612413788766d7c98135                                                                                                                                    0.0s
 => => naming to docker.io/pytorch/torchserve:latest-cpu                                                                                                                                                                        0.0s
~/projects/serve/docker(master)$ docker run -it pytorch/torchserve:latest-cpu python --version
Python 2.7.18
~/projects/serve/docker(master)$ docker run -it pytorch/torchserve:latest-cpu python3 --version
Python 2.7.18
~/projects/serve/docker(master)$ 
```
- [x] Python version after fix
```
~/projects/serve/docker(fix_docker_python_version)$ ./build_image.sh -py 3.10 
Valid python version
[+] Building 2.3s (23/23) FINISHED                                                                                                                                                                                                   
 => [internal] load build definition from Dockerfile                                                                                                                                                                            0.0s
 => => transferring dockerfile: 5.36kB                                                                                                                                                                                          0.0s
 => [internal] load .dockerignore                                                                                                                                                                                               0.0s
 => => transferring context: 2B                                                                                                                                                                                                 0.0s
 => resolve image config for docker.io/docker/dockerfile:experimental                                                                                                                                                           1.2s
 => CACHED docker-image://docker.io/docker/dockerfile:experimental@sha256:600e5c62eedff338b3f7a0850beb7c05866e0ef27b2d2e8c02aa468e78496ff5                                                                                      0.0s
 => [internal] load metadata for docker.io/library/ubuntu:20.04                                                                                                                                                                 0.9s
 => [compile-image 1/8] FROM docker.io/library/ubuntu:20.04@sha256:24a0df437301598d1a4b62ddf59fa0ed2969150d70d748c84225e6501e9c36b9                                                                                             0.0s
 => [internal] load build context                                                                                                                                                                                               0.0s
 => => transferring context: 80B                                                                                                                                                                                                0.0s
 => CACHED [runtime-image 2/9] RUN --mount=type=cache,target=/var/cache/apt     apt-get update &&     apt-get upgrade -y &&     apt-get install software-properties-common -y &&     add-apt-repository -y ppa:deadsnakes/ppa   0.0s
 => CACHED [runtime-image 3/9] RUN useradd -m model-server     && mkdir -p /home/model-server/tmp                                                                                                                               0.0s
 => CACHED [compile-image 2/8] RUN --mount=type=cache,id=apt-dev,target=/var/cache/apt     apt-get update &&     apt-get upgrade -y &&     apt-get install software-properties-common -y &&     add-apt-repository -y ppa:dead  0.0s
 => CACHED [compile-image 3/8] RUN update-alternatives --install /usr/bin/python python /usr/bin/python3.10 1     && update-alternatives --install /usr/local/bin/pip pip /usr/local/bin/pip3 1                                 0.0s
 => CACHED [compile-image 4/8] RUN python3.10 -m venv /home/venv                                                                                                                                                                0.0s
 => CACHED [compile-image 5/8] RUN python -m pip install -U pip setuptools                                                                                                                                                      0.0s
 => CACHED [compile-image 6/8] RUN export USE_CUDA=1                                                                                                                                                                            0.0s
 => CACHED [compile-image 7/8] RUN TORCH_VER=$(curl --silent --location https://pypi.org/pypi/torch/json | python -c "import sys, json, pkg_resources; releases = json.load(sys.stdin)['releases']; print(sorted(releases, key  0.0s
 => CACHED [compile-image 8/8] RUN python -m pip install -U setuptools && python -m pip install --no-cache-dir captum torchtext torchserve torch-model-archiver                                                                 0.0s
 => CACHED [runtime-image 4/9] COPY --chown=model-server --from=compile-image /home/venv /home/venv                                                                                                                             0.0s
 => CACHED [runtime-image 5/9] COPY dockerd-entrypoint.sh /usr/local/bin/dockerd-entrypoint.sh                                                                                                                                  0.0s
 => CACHED [runtime-image 6/9] RUN chmod +x /usr/local/bin/dockerd-entrypoint.sh     && chown -R model-server /home/model-server                                                                                                0.0s
 => CACHED [runtime-image 7/9] COPY config.properties /home/model-server/config.properties                                                                                                                                      0.0s
 => CACHED [runtime-image 8/9] RUN mkdir /home/model-server/model-store && chown -R model-server /home/model-server/model-store                                                                                                 0.0s
 => CACHED [runtime-image 9/9] WORKDIR /home/model-server                                                                                                                                                                       0.0s
 => exporting to image                                                                                                                                                                                                          0.0s
 => => exporting layers                                                                                                                                                                                                         0.0s
 => => writing image sha256:a7b59a9682a73073e1aec38710ebcfbef70118e4bd0fbd36082f5a9c97ad11eb                                                                                                                                    0.0s
 => => naming to docker.io/pytorch/torchserve:latest-cpu                                                                                                                                                                        0.0s
~/projects/serve/docker(fix_docker_python_version)$ docker run -it pytorch/torchserve:latest-cpu python --version
Python 3.10.10
~/projects/serve/docker(fix_docker_python_version)$ docker run -it pytorch/torchserve:latest-cpu python3 --version
Python 3.10.10
~/projects/serve/docker(fix_docker_python_version)$ 
```

## Checklist:

- [x] Did you have fun?
- [x] Have you added tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?

### CODE REVIEW

**Comment by agunapal** [comment]
@fabridamicelli Thanks for reporting! It would be great if you can contribute to docker CI. Currently, we don't have one

**Comment by fabridamicelli** [comment]
> @fabridamicelli Thanks for reporting! It would be great if you can contribute to docker CI. Currently, we don't have one

@agunapal Thanks for your reply!
ok, I will try to draft something simple for CI and we can build from there.

**Review by agunapal — APPROVED** [review]

**Review by msaroufim — APPROVED** [review]

**Comment by fabridamicelli** [comment]
@agunapal I am facing some problems with the fixed image (which uses python3.9) as default. I am working to fix the problem. Please wait until I fix that to merge into main because that seems to break some builds. I'll ping you when it's ready on my side

**Comment by fabridamicelli** [comment]
@agunapal @msaroufim  I took a closer look at and refactored the `Dockerfile` and now it's ready for you to review it in my opinion.

Here's the updated summary of the changes and the rationale behind them:
- **delete unnecessary pip bootstrapping**: running apt-get install python$PYTHON_VERSION-dev is enough. Also this pip was not being used, since the venv is set up later on (as evidenced by the step `python -m pip install -U pip`).
- **delete duplicate pip upgrade**: command python -m pip install -U pip was repeated.
- **delete update-alternatives magic**: it is not necessary because setting ENV PATH="/home/venv/bin:$PATH" kind of "moves the pointer" to use `/home/venv/pip`, so the right `python` and `pip` versions are used after that point (I added a comment on that to the Dockerfile)
- **added a bash script** (`/docker/build_image_and_check_container.sh`) to check that:
  - image builds fine and container runs with a healthy status
  - python version in the container is indeed the correct one (ie, passed to `./build_image.sh`)
 
I ran the script (eg `./docker/build_image_and_check_container.sh 3.9`) for python 3.8, 3.9 and 3.10 locally and everything works fine.

**Possible next steps**
- Integrate the check script (or something similar) in CI to make sure that future changes in `Dockerfile` don't break things. 
- Push extra images to dockerhub with different python versions, eg `pytorch/torchserve:0.7.1-cpu-py3.10`

I'd be happy to contribute along those lines too if you folks think it'd be useful (?)

**Comment by agunapal** [comment]
@fabridamicelli 

Awesome This is great!

Could you please include the logs of this example here to make sure nothing is broken
https://github.com/pytorch/serve/blob/master/examples/image_classifier/mnist/Docker.md

Please feel free to send a PR for CI, github action. You can refer to one of the existing ones.

It would be great if the test I shared above can be included in this.

**Comment by fabridamicelli** [comment]
> @fabridamicelli
> 
> Awesome This is great!
> 
> Could you please include the logs of this example here to make sure nothing is broken https://github.com/pytorch/serve/blob/master/examples/image_classifier/mnist/Docker.md
> 
> Please feel free to send a PR for CI, github action. You can refer to one of the existing ones.
> 
> It would be great if the test I shared above can be included in this.

@agunapal 
That sounds good, thank you for pointing me to the relevant example.

I am working on a little CI workflow to have directly integrated in this PR that checks the image/container so that we don't have to rely on the printed out logs :)
It shouldn't take that long to have it working, I'll ping you when it's there!

**Comment by fabridamicelli** [comment]
@agunapal 
I managed to set up a CI workflow to test the docker image produced by Dockerfile for all three supported python versions:
- build_image runs successfully
- container from image has "Healthy" status
- python version in container is indeed the desired (passed to ./build_image.sh -py)
- generate correct predictions from model running in container (MNIST example data that you pointed me to): Notice that (in contrast to the example) all the steps (model archiving, etc) happen inside the container and the example files are mounted directly which allows us to test it as a stand-alone example, completely isolated and without need of manual steps. 
	
To see the workflow in action already you can take a look at [this run](https://github.com/fabridamicelli/torchserve-docker/actions/runs/4583998597/jobs/8095203319) on a repo I set up for testing where I pull this branch and let the workflow run.

I think that with that we could wrap up this PR before it gets too large. But as potential next steps I see: 
- extending tests to build images with other relevant parameters
- Push extra images to dockerhub with different python versions, eg `pytorch/torchserve:0.7.1-cpu-py3.10` (would you folks be interested in this?)

**Review by agunapal — APPROVED** [review]
LGTM

**Review by agunapal — APPROVED** [review]

**Comment by agunapal** [comment]
@fabridamicelli  Thanks! This is great. Please feel free to send new PRs for adding more tests to Docker CI

File: .github/workflows/test-docker-container.yml
Comment (msaroufim) [inline .github/workflows/test-docker-container.yml:38]: neat!

**Review by msaroufim — APPROVED** [review]
This is great @fabridamicelli - we had chatted on the team about doing something like this for a while but thank you for going ahead and doing it

Since this is a set of tests and they're not required signals please make sure they're all green before merge

File: .github/workflows/test-docker-container.yml
Comment (agunapal) [inline .github/workflows/test-docker-container.yml:33]: @fabridamicelli Can you please change the working directory here. The run fails here

**Review by agunapal — CHANGES_REQUESTED** [review]

File: .github/workflows/test-docker-container.yml
Comment (fabridamicelli) [reply .github/workflows/test-docker-container.yml:33]: yes! I just pushed the fix, so you can re-trigger the CI and it should (hopefully) work :)

**Review by agunapal — APPROVED** [review]

**Pull request merged**
